### ISSUE TITLE
Remove usage of hotshot in twistd

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5137]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#5137|
|Type|task|
|Created|2011-06-06 17:22:13Z|
|Branch|https://github.com/twisted/twisted/tree/rip-in-peace-hotshot-5137|

Twisted emits the following deprecation warning under Python 2.7 with the `-3` option:

```
The 'hotshot' module is not supported in 3.x, use the 'profile' module instead.
```

The only remaining usage of hotshot is in ```twisted.application``` (twistd). This usage should be deprecated.

The [Python docs](http://docs.python.org/library/hotshot.html) recommend using `cProfile` instead.

<details><summary>Searchable metadata</summary>

```
trac-id__5137 5137
type__task task
reporter__thijs thijs
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_rip_in_peace_hotshot_5137 branches/rip-in-peace-hotshot-5137
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1307380933000000 1307380933000000
changetime__1437294340030299 1437294340030299
version__None None
owner__hawkowl hawkowl
cc__thijs cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

The usage in `twisted/application/app.py` and `twisted/test/test_twistd.py` are to support the option of profiling with `hotshot`.  They handle import errors and behave sanely when the imports fail, so they aren't going to cause a major problem on Python 3.x.  Since hotshot is not a very good profiler, we may want to deprecate this command line option anyway.

`profiled`, the cause of the usage in `twisted/trial/util.py`, should be deprecated and `trial` should share the same profiling option/implementation code as is used by `twistd`.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

Replying to [thijs](#5137):
> The [Python docs](http://docs.python.org/library/hotshot.html) recommend using `cProfile` instead.

See also [#3356](https://github.com/twisted/twisted/issues/3356).

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_3"></a>|@zooko commented|
|-|-|

I'm sorry, but I'm a bit confused by JP's comments in comment:1. Is this the ticket for me? I want to make it so that when I tell people to do ```tahoe start --profile```, which results in [--profile=profiling_results.prof --savestats](http://tahoe-lafs.org/trac/tahoe-lafs/browser/trunk/src/allmydata/scripts/startstop_node.py?annotate=blame&rev=5074#L71) being passed to ```twistd```, that they no longer use ```hotshot```.

The reason I care is that when they use ```hotshot``` then they send me data files which, when I try to examine them do this:
```
  File "/usr/lib/python2.7/hotshot/stats.py", line 54, in load
    assert not self._stack
AssertionError
```

Which is a bug [closed as wontfix](http://bugs.python.org/issue900092) by the Python developers.

So I'm not currently concerned about ```trial```, only about ```twistd```. Am I at the right ticket?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Why do you translate `tahoe start --profile` into `[twistd] --profile=profiling_results.prof --savestats`?  Instead, you should translate it to `[twistd] --profile=profiling_results.prof --savestats --profiler=cprofile`.  See the twistd --help output or the man page for usage information.

This ticket is not about the default profiler.  It's about not using hotshot in Twisted except where we must for backwards compatibility; ie, it is for removing the _direct_ usage of hotshot from trial, in favor of letting trial and twistd share twistd's existing hotshot support code.  Alternatively (or additionally), we may _also_ want to deprecate hotshot support entirely and then remove it.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_5"></a>|@zooko commented|
|-|-|

Replying to [exarkun](#note_4):
> Why do you translate `tahoe start --profile` into `[twistd] --profile=profiling_results.prof --savestats`?  Instead, you should translate it to `[twistd] --profile=profiling_results.prof --savestats --profiler=cprofile`.

The reasoning was to let Twisted choose the profiler to use rather than to specify which one it should use. Should there be a Twisted ticket to change what the default profiler is when the user asks for profiling but doesn't specify a profiler?

In the meantime, we'll do as you suggest, but in the future if "cProfile" stops working (possibly because we've moved to PyPy?), it would be nice if we could ask Twisted to do profiling without having to do our own research and learn which profilers are broken and unsupported. :-)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(In [45189]) Branching to deprecate-hotshot-5137.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

The only usage of hotshot is now in twistd, so this ticket deprecates the Hotshot runner and makes `profile` the default.

Builders spun, looking good so far, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for working on this!

I think that we should update the ticket description to reflect what is done in the final branch.

I think that we should still keep hotspot as the default profiler to preserve backward compatibility.
Once hotspot is removed, we can change the default profiler.

What do you think? You are the release manager, so feel free to not follow my comments :)

Please review my comments and either merge as it is now or come back with latest changes.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Hi Adi,

I'll update the ticket.

I've asked Glyph for his opinion, since I'm a bit split over it. Might email the mailing list later.

Of course, I'm just a ticket submitter right now, my RM gives me no privs to ignore review comments without reason ;)

I'll think on this and see what needs doing. Thanks again for the review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

When you say  "Twisted emits the following deprecation warning under Python 2.7 with the -3 option", what exactly makes "twisted" emit this?  `twistd --help`?  `twistd --profile <something>`?  `import twisted`?

I'm inclined to say we should compatibility-exception our way out of this, and maybe even edit `CompatibilityPolicy` to say specifically that things no longer supported by upstream devs can be exceptions automatically; `hotshot` is clearly unmaintained and should not be used.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

The original ticket might have either been the second (--profile) or the test suite -- py3k-warnings outputs this.

```
/buildslave/python-3k-warnings/Twisted/twisted/test/test_twistd.py:67: DeprecationWarning: The 'hotshot' module is not supported in 3.x, use the 'profile' module instead.
```

I'll broadcast this out to the mailing list.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

(In [45223]) Branching to rip-in-peace-hotshot-5137.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl removed owner|
|-|-|

Let's just remove it entirely.

As this is a backwards incompatible change, I'm also sending something to the mailing list.

Builders looking green, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

Changes look good. Maybe the news segment should mention that hotstop was removed as it was already broken.

This is also code designed for development so I think that we should more relaxed with the deprecation process.

Feel free to merge it.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_16"></a>|@moshez commented|
|-|-|

I think removing this entirely is fine.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_17"></a>|@moshez set owner to @hawkowl|
|-|-|

Looks good to me -- I doubt anyone is using --profile=hotshot in production and would be hurt by this.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45273]) Merge rip-in-peace-hotshot-5137: Remove hotshot from twistd

Author: hawkowl
Reviewers: adiroiban, moshez
Fixes: [#5137](https://github.com/twisted/twisted/issues/5137)
